incorrect parens around a few 'extern char*' decls (most importantly: tparm, tgoto, which cause a core dump or compile errors under certain environments, e.g. CentOS7) no longer use -u on jmake since it causes a spurious failure in some situations

markmoraes opened the pull request:
incorrect parens around a few 'extern char*' decls (most importantly: tparm,
    tgoto, which cause a core dump or compile errors under certain environments,
    e.g. CentOS7)
    no longer use -u on jmake since it causes a spurious failure in some situations

markmoraes committed:
incorrect parens around a few 'extern char*' decls (most importantly: tparm,
tgoto, which cause a core dump or compile errors under certain environments,
e.g. CentOS7)
no longer use -u on jmake since it causes a spurious failure in some situations

vixie commented:
I think the parents are unnecessary in the fix, we can simply remove them.


Sent from Workspace ONE Boxer

markmoraes commented:
We have parens around lots of the extern declarations (though it seems not all) to avoid complaints if any of them are macros (was that only pre-stdc?  I would guess only a few are macros these days -- maybe putchar or suchlike).  More broadly, should we even be defining things that stdc promises are macros if we now draw c2x as a baseline: most of these externs like getcwd, etc were either for machines with very deficient headers without some of these being defined in any sane/standard place we could count on, or where the systems did not have prototypes, so we were providing them if they were compiled with a stdc compiler like gcc.  Should we just remove all the parens and superfluous externs as part of c2x?

vixie commented:
<<Should we just remove all the parens and superfluous externs as part of c2x?>>


Yes.


Sent from Workspace ONE Boxer

markmoraes committed:
undo some editing damage in jtc.c (only affected the standalone test code)

removed all declarations that would appear in C89 or POSIX.1 standard headers
(I think POSIX.1 1988 ought to be fine, though hard to say since all machines
I can test are far more modern and conform to 2001)

restored JSSIZE_T so Windows builds again.

Tweaks to testbuild.sh to test mingw and pipeprocs.

markmoraes committed:
removed one more unnecessary extern for _exit()

markmoraes committed:
updated sysdep.txt to delete obsolete config options

markmoraes committed:
remove more ancient/pre-POSIX/untested stuff: TERMIO, SGTTY,
USE_GETWD/USE_PWD, WAIT3 or BSD_WAIT.

markmoraes commented:
tested this on Windows (mingw and msvc) and a bunch of different rarely-used permutations on Linux (SMALL, PIPEPROCS), did a bit of cleanup of sysdep.txt and sysdep.h as well.  removed a lot of pre-POSIX ifdefs (pretty much untested code paths for two decades)

vixie commented on unix.c:
```diff
@@ -344,7 +215,7 @@ ttysetattr(jbool n)	/* `n' is also used as subscript! */
 		static struct stat	tt_stat;
 # if !defined(USE_FSTAT) || !defined(USE_FCHMOD)
 		static char	*tt_name = NULL;	/* name of the control tty */
-		extern char	*ttyname(int);		/* for systems w/o fstat */
```
what is the purpose of this change?

vixie replied:
what i see in the diff is replacement of *ttyname() def'n to be *(ttyname)(). in other words adding back some parens. i think this def'n can simply be deleted.

vixie reviewed (commented):
there are two parenthesis-level changes that i've asked questions about.

vixie commented on recover.c:
```diff
@@ -74,7 +72,7 @@ char	*argv[];
  * it for OSF.
  */
 #ifndef	_OSF_SOURCE
-extern FILE	*popen(const char *, const char *);
```
what i see in the diff is replacement of *popen() def'n to be *(popen)(). in other words adding back some parens. i think this def'n can simply be deleted.

vixie commented:
yes, let's delete those definitions (popen and ttyname). but the PR as written simply restores the parens in those two cases.

markmoraes committed:
delete popen declaration, who has OSF/1 and I very much
doubt c2x Jove compiles/works on it anyway, old systems need
to stay with 4.17.

markmoraes commented:
Apologies, I was misreading the excerpt above instead of looking at the actual PR branch. 

I'm fine deleting popen(), of course, I updated and pushed to the PR to do that.  I think the reason I kept it was the comment above, that it isn't part of POSIX.1 but rather .2.  That said, I'm prepared to toss OSF/1 overboard with the raft of ancient systems we can't/won't support any longer.

I'm less sure about deleting the ttyname declaration because that seems to make the ifdef that it is in pointless.  I would argue that systems without fstat and fchmod are unlikely to have a ttyname declaration in the right header file.  So rather than delete that declaration, I see two choices:
1. nuke the ifdef branches for systems that do not have USE_FSTAT and USE_FCHMOD (this biff code in unix.c is the only use of those ifdefs and of fstat/fchmod)
2. nuke BIFF support entirely, bit more drastic/wider, but surely that biff logic (which chmod -x `ttyname` to suppress the ancient/extinct in.comsatd writing to the terminal and messing up the display) can't be in use anymore -- I haven't invoked biff myself since 1990 at least, perhaps even before that, comsatd made no sense with networked mail spool or imap/pop/webmail/apps.  It's a tiny amount of code, but feels like an improvement overall to eliminate.

markmoraes commented:
I propose I use separate PRs in the c2x branch to nuke biff (and there's certainly a lot of other code I see that is painfully obsolete and I would really like to kill -- old Mac support, the unused xjove/* pertaining to jovetool/xjovetool (except for xjove/mousemsg.h which I propose to move upto the main dir since it is needed for xterm mouse support, which is at least avail on modern systems; I'm not going looking for xview for the rest of xjovetool) and I'm sure I'll run across some other ancientness (7bittedness?  USE_FSYNC, anything non-POSIX for sigs, procs, ptys), each of which is probably best done as a separate PR for better testing and mistake isolation.

markmoraes committed:
remove ttyname declaration (systems that want biff support
had either better have fstat+fchmod or declare ttyname in
unistd.h)

markmoraes commented:
Yup, already pushed that one
